docs/book/tutorial/page-3.md

```diff
@@ -0,0 +1,736 @@
+		})
+		for i, job := range successfulJobs {
+			if err := r.Delete(ctx, job); err != nil {
+				log.Error(err, "unable to delete old successful job", "job", job)
+			}
+			if int32(i) >= *cronJob.Spec.SuccessfulJobsHistoryLimit {
+				break
+			}
+		}
+	}
+
+```
+{%% endmethod %%}
+
+{% method %}
+
+Then, we'll check and see if any jobs should be running.  If the CronJob is suspended,
+we shouldn't do anything.
+
+```go
+
+
+	// get on with the actual reconciling
+
+	// check to see if we're supposed to do anything
+	// (user might've wanted a pause to putz with stuff)
+	if cronJob.Spec.Suspend != nil && *cronJob.Spec.Suspend {
+		log.V(1).Info("cronjob suspended, skipping")
+		return ctrl.Result{}, nil
+	}
+
+```
+{%% endmethod %%}
+
+{% method %}
+
+Otherwise, we'll need to calculate any runs that are past due (in which case we
+might need to run a job), as well as the next run, so that we know how long we
+need to wait until checking again.
+
+
+```go
+
+
+	// figure out the next times that we need to create
+	// jobs at (or anything we missed).
+	missedRuns, nextRun, err := getNextSchedule(&cronJob, r.Now())
+	if err != nil {
+		log.Error(err, "unable to figure out CronJob schedule")
+		// we don't really care about requeuing until we get an update that fixes the schedule,
+		// so don't return an error
+		return ctrl.Result{}, nil
+	}
+
+```
+<a name="return-from-func-getNextSchedule"></a>
+[func getNextSchedule](#jump-to-func-getNextSchedule)
+{%% endmethod %%}
+
+{% method %}
+
+We'll prep our eventual request to requeue until the next job, and then figure
+out if we actually need to run.  If we've missed a run, and we're still within
+the deadline to start it, we'll need to run a job.
+
+```go
+
+
+	scheduledResult := ctrl.Result{RequeueAfter: nextRun.Sub(r.Now())} // save this so we can re-use it elsewhere
+	log = log.WithValues("now", r.Now(), "next run", nextRun)
+
+	if len(missedRuns) == 0 {
+		log.V(1).Info("no upcoming scheduled times, sleeping until next")
+		return scheduledResult, nil
+	}
+
+	if len(missedRuns) > 1 {
+		log.V(1).Info("multiple missed scheduled runs, on running most recent")
+	}
+
+
+	// make sure we're not too late to start the run
+	lastMissedRun := missedRuns[len(missedRuns)-1]
+	log = log.WithValues("current run", lastMissedRun)
+	tooLate := false
+	if cronJob.Spec.StartingDeadlineSeconds != nil {
+		tooLate = lastMissedRun.Add(time.Duration(*cronJob.Spec.StartingDeadlineSeconds)*time.Second).Before(r.Now())
+	}
+	if tooLate {
+		log.V(1).Info("missed starting deadline for last run, sleeping till next")
+		// TODO(directxman12): events
+		return scheduledResult, nil
+	}
+
+```
+{%% endmethod %%}
+
+{% method %}
+
+If we actually have to run a job, we'll need to either wait till existing ones finish,
+replace the existing ones, or just add new ones.  If our information is out of date due
+to cache delay, we'll get a requeue when we get up-to-date information.
```

Comment:
> If our information is out of date due to cache delay, we'll get a requeue when we get up-to-date information.

It's not very clear which piece of code is actually dealing with this case. I'd like to discuss about this more in person.

Reply:
Discussed with Solly offline. He is going to rephrase this piece.